gitAndTools.git-remote-hg: 0.2 -> 1.0.0

erosennin opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

git-remote-hg-0.2 is broken with recent versions of Mercurial:

```
$ hg init test-hg
$ cd test-hg
$ touch README
$ hg add
$ hg commit -m 'Add readme'
$ cd ..
$ git clone hg::test-hg test-git
Cloning into 'test-git'...
Traceback (most recent call last):
  File "/nix/store/5m6xvmcqq5j248i90hnrga7pal6wm3yr-git-remote-hg-0.2-e716a9e1a9e460a45663694ba4e9e8894a8452b2/bin/.git-remote-hg-wrapped", line 1333, in <module>
    sys.exit(main(sys.argv))
  File "/nix/store/5m6xvmcqq5j248i90hnrga7pal6wm3yr-git-remote-hg-0.2-e716a9e1a9e460a45663694ba4e9e8894a8452b2/bin/.git-remote-hg-wrapped", line 1317, in main
    do_import(parser)
  File "/nix/store/5m6xvmcqq5j248i90hnrga7pal6wm3yr-git-remote-hg-0.2-e716a9e1a9e460a45663694ba4e9e8894a8452b2/bin/.git-remote-hg-wrapped", line 729, in do_import
    export_bookmark(repo, bmark)
  File "/nix/store/5m6xvmcqq5j248i90hnrga7pal6wm3yr-git-remote-hg-0.2-e716a9e1a9e460a45663694ba4e9e8894a8452b2/bin/.git-remote-hg-wrapped", line 613, in export_bookmark
    export_ref(repo, bmark, 'bookmarks', head)
  File "/nix/store/5m6xvmcqq5j248i90hnrga7pal6wm3yr-git-remote-hg-0.2-e716a9e1a9e460a45663694ba4e9e8894a8452b2/bin/.git-remote-hg-wrapped", line 486, in export_ref
    revs = gitrange(repo, tip, head)
  File "/nix/store/5m6xvmcqq5j248i90hnrga7pal6wm3yr-git-remote-hg-0.2-e716a9e1a9e460a45663694ba4e9e8894a8452b2/bin/.git-remote-hg-wrapped", line 453, in gitrange
    pending = set([int(b)])
TypeError: int() argument must be a string or a number, not 'changectx'
fatal: stream ends early
fast-import: dumping crash report to /tmp/test-git/.git/fast_import_crash_16959
fatal: error while running fast-import
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

erosennin commented:
Should probably be also backported to 19.03.

worldofpeace commented on pkgs/applications/version-management/git-and-tools/git-remote-hg/default.nix:
```diff
@@ -1,33 +1,32 @@
-  rev = "e716a9e1a9e460a45663694ba4e9e8894a8452b2";
-  version = "0.2-${rev}";
+  version = "1.0.0";
   name = "git-remote-hg-${version}";
```
```suggestion
  pname = "git-remote-hg";
```

Please reuse the attribute where it makes sense

> stdenv.mkDerivation may now be passed pname and version attributes instead of name, and will now automatically construct the package name by joining them with a dash.

worldofpeace commented on pkgs/applications/version-management/git-and-tools/git-remote-hg/default.nix:
```diff
@@ -1,33 +1,32 @@
 
   meta = with stdenv.lib; {
-    homepage = https://github.com/felipec/git-remote-hg;
+    homepage = https://github.com/mnauw/git-remote-hg;
```
I'm seeing lots of url changes here.

There's
* fingolfin
* felipec
* mnauw

Which is the proper source, when did these changes happen? (forks?).

erosennin replied:
The original repo is [mnauw/git-remote-hg](https://github.com/mnauw/git-remote-hg). Looks like it has been dormant for a while, so a couple of forks appeared, but now it seems to be active again. [Debian uses it, too.](https://salsa.debian.org/debian/git-remote-hg/blob/master/setup.py#L45)

worldofpeace commented on pkgs/applications/version-management/git-and-tools/git-remote-hg/default.nix:
```diff
@@ -1,33 +1,32 @@
   meta = with stdenv.lib; {
-    homepage = https://github.com/felipec/git-remote-hg;
+    homepage = https://github.com/mnauw/git-remote-hg;
     description = "Semi-official Mercurial bridge from Git project, once installed, it allows you to clone, fetch and push to and from Mercurial repositories as if they were Git ones";
```
Can we shorten this?
It's pretty long for `description` which is meant to be short and to the point.

worldofpeace commented on pkgs/applications/version-management/git-and-tools/git-remote-hg/default.nix:
```diff
@@ -1,33 +1,32 @@
 
   doCheck = false;
 
   installFlags = "HOME=\${out} install-doc";
```
```suggestion
  installFlags = [ "HOME=${placeholder ''out''}"  "install-doc" ];
```

worldofpeace commented on pkgs/applications/version-management/git-and-tools/git-remote-hg/default.nix:
```diff
@@ -1,33 +1,32 @@
+  src = fetchFromGitHub {
+    owner = "mnauw";
+    repo = "git-remote-hg";
+    rev = "765f9ae2875a2ef4439470fb4cf4adefdda2ed67";
```
```suggestion
    rev = "v${version}";
```

There's a tag.

worldofpeace commented on pkgs/applications/version-management/git-and-tools/git-remote-hg/default.nix:
```diff
@@ -1,33 +1,32 @@
-    sha256 = "0cmlfdxfabrs3x10mfjfap8wz67s8xk2pjn2wlcj9k2v84gji60m";
+  src = fetchFromGitHub {
+    owner = "mnauw";
+    repo = "git-remote-hg";
```
```suggestion
    repo = pname;
```

worldofpeace reviewed (changes requested):
See above minor comments.

This appears to be a python application https://github.com/mnauw/git-remote-hg/blob/master/setup.py.
So we should be using `buildPythonApplication` and fetching with `fetchPypi`.
##### https://pypi.org/project/git-remote-hg/

You should also document the upstream source changes in the commit msg.

worldofpeace commented on pkgs/applications/version-management/git-and-tools/git-remote-hg/default.nix:
```diff
@@ -22,7 +22,7 @@ stdenv.mkDerivation rec {
   postInstall = ''
     wrapProgram $out/bin/git-remote-hg \
-      --prefix PYTHONPATH : "$(echo ${mercurial}/lib/python*/site-packages):$(echo ${mercurial.python}/lib/python*/site-packages)${stdenv.lib.concatMapStrings (x: ":$(echo ${x}/lib/python*/site-packages)") mercurial.pythonPackages or []}"
+      --set PYTHONPATH "$(echo ${mercurial}/lib/python*/site-packages):$(echo ${mercurial.python}/lib/python*/site-packages)${stdenv.lib.concatMapStrings (x: ":$(echo ${x}/lib/python*/site-packages)") mercurial.pythonPackages or []}"
```
`buildPython*` should be able to wrap this when you put the runtime deps in `propagatedBuildInputs`.

erosennin replied:
The problem is, `setup.py` is currently broken, because it tries to fetch the version git a git tag and fails. We can patch it though.

erosennin replied:
> buildPython* should be able to wrap this when you put the runtime deps in propagatedBuildInputs.

This package depends on `mercurial`, which is built with `buildPythonApplication`. Will it still work?

worldofpeace replied:
> The problem is, `setup.py` is currently broken, because [it tries to fetch the version from a git tag](https://github.com/mnauw/git-remote-hg/blob/f050de1bcc3f37f5572035a63c85aba60f7b9088/setup.py#L9) and fails. We can patch it though.

Yeah they're many packages that take a much better approach that they could have used as an example. Patching and submitting it upstream would be preferred. 

But if it's too much of a hassle, you can use `format = "other";` and just use the `Makefile` as normally.

> https://nixos.org/nixpkgs/manual/#python  9.14.2.2.1.1. buildPythonPackage parameters


> This package depends on mercurial, which is built with buildPythonApplication. Will it still work?

Think so.

erosennin commented:
Wow, thanks for such a thorough review. :)

erosennin committed:
gitAndTools.git-remote-hg: 0.2-e716a9e1a9e460a45663694ba4e9e8894a8452b2 -> 1.0.0

Changed the upstream repo to https://github.com/mnauw/git-remote-hg:

- looks the most alive at the moment, the last commit is from March 2019;
- https://pypi.org/project/git-remote-hg/ points to this repo as well;
- Debian uses it too:
  https://salsa.debian.org/debian/git-remote-hg/blob/e769897119850979d590b758791b7d7b36db3c41/setup.py#L45

erosennin committed:
gitAndTools.git-remote-hg: refactor

Use buildPythonApplication, organize build inputs.

erosennin commented:
Refectored to use `buildPythonApplication`.

worldofpeace commented on pkgs/applications/version-management/git-and-tools/git-remote-hg/default.nix:
```diff
@@ -1,34 +1,37 @@
+      --replace "version = process.communicate()[0].strip()" 'version = "v${version}"'
+  '';
 
   postInstall = ''
```
You can drop this and use
```
installFlags = [
  "HOME=${placeholder '''out''}" 
  "install-doc"
];
````

like before.
```

erosennin replied:
?

If I understand correctly, `buildPythonApplication` runs `pip install ...` in `installPhase` and passes `installFlags` to `pip install`. Passing `HOME=$out install-doc` to `pip install` does not make much sense. :) What we want instead is to install the Python package with `pip install`, then install the manpage with `make HOME=$out install-doc`.

worldofpeace replied:
Oops, you can ignore this comment because it wasn't meant to be posted :innocent:

worldofpeace commented on pkgs/applications/version-management/git-and-tools/git-remote-hg/default.nix:
```diff
@@ -1,34 +1,37 @@
+  preConfigure = ''
+    substituteInPlace setup.py \
+      --replace "process = subprocess.Popen(cmd, stdout=subprocess.PIPE)" "" \
+      --replace "version = process.communicate()[0].strip()" 'version = "v${version}"'
```
I you're going to do this, can you use a patch in  `patches`?

erosennin replied:
OK, done.

worldofpeace commented on pkgs/applications/version-management/git-and-tools/git-remote-hg/default.nix:
```diff
@@ -1,34 +1,37 @@
+  nativeBuildInputs = [ asciidoc xmlto docbook_xsl docbook_xml_dtd_45 libxslt libxml2 ];
+  propagatedBuildInputs = [ mercurial ];
 
   doCheck = false;
```
Has tests https://github.com/mnauw/git-remote-hg/blob/master/Makefile#L10

Maybe try to enable them.

erosennin replied:
Looks like the tests are going to be problematic. They mostly pass but one or two of them fail. What's worse, the failures are different with different versions of Git and Mercurial. I'll have to look into it.

worldofpeace commented on pkgs/applications/version-management/git-and-tools/git-remote-hg/default.nix:
```diff
@@ -1,34 +1,35 @@
 
-  installFlags = "HOME=\${out} install-doc";
+  patches = [
+    ./patches/./no-dynamic-git-version.patch
```
```suggestion
    ./patches/no-dynamic-git-version.patch
```

erosennin commented:
Already done in #57730.
